Initial implementation

dodsky opened the pull request.

Commit:
Initial implementation

Commit:
Refactoring after initial usage, added more test coverage

Commit:
Added more tests, revised some configuration options

Commit:
Refactored config to enable promises as return values from config functions

Commit:
Added license, fixed version

vedrani commented on src/helpers/tokenFormatter.js:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  return token;
+}
```
This should maybe be default way of handling authorization, but we should also allow to override it in config from Builder.

vedrani commented on src/helpers/tokenFormatter.js:
```diff
@@ -0,0 +1,25 @@
+export function formatBearer(token) {
```
This is only for tests? If so move it to /test/helpers

vedrani commented on src/const.js:
```diff
@@ -0,0 +1,11 @@
+export const ENVIRONMENT_IS_REACT_NATIVE = typeof navigator === 'object' && navigator.product === 'ReactNative';
+export const ENVIRONMENT_IS_NODE = typeof process === 'object' && typeof require === 'function';
+export const ENVIRONMENT_IS_WEB = typeof window === 'object';
+export const ENVIRONMENT_IS_WORKER = typeof importScripts === 'function';
```
This should be changed into functions with camelCase and moved maybe to service/environment.js

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+      return fetchInterceptor(fetch, ...args);
+    };
+  })(env.fetch);
+}
```
This function with above check of environment should be encapsulated into `environment` that would be called in index.js like environment.init(); only once upon index.js load. Names are just suggestions.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  refreshToken: null,
+};
+
+let refreshAccessTokenPromise = null;
```
All of above `let` variables should be changed into `const` and moved into encapsulated parts of lib that use them.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+import { parseBearer } from './helpers/tokenFormatter';
```
This file should be divided into separate parts/classes. Probably index.js that just configures environment and classes/services:

- Interceptor - root object that Builder initializes passing config (not singleton), initializes AccessTokenProvider in ctr based on received config, responsible for intercepting requests, checking is responsible for managing authorization, knows how to create request, know how to create inner promise.  Has main method that intercepts in which is simple flow that calls inner methods, AccessTokenProvider methods.
- AccessTokenProvider - not singleton, has request token and is responsible for fetching new access token, exposes methods for simple check is token ok, should it be fetched, fetching new token that returns promise for chaining, maybe even know to add token to header
- Environment/Bootstraper,...
- services...

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+      .then(response => {
+        refreshAccessTokenPromise = null;
+
+        if (response.status !== STATUS_OK) {
```
Turn into call to service/http.js isResponseOk

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  })(env.fetch);
+}
+
+function runRefreshTokenPromise() {
```
Part of AccessTokenProvider, something like `renew()` that returns promise

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+} from './const';
+
+let config = {
+  prepareRefreshTokenRequest: null,
```
createAccessTokenRequest

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  shouldIntercept: null,
+  shouldInvalidateAccessToken: null,
+  getAccessTokenFromResponse: null,
+  setRequestAuthorization: null,
```
setAuthorizationHeader?

dodsky replied:
Thought about it, should be more general, what if authorization goes thru query param or any other place? It isn't by the spec, but we probably don't want to be that restrictive, the goal of this method is to authorize request in any way specified. I'm renaming it to `authorizeRequest`.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  shouldInvalidateAccessToken: null,
+  getAccessTokenFromResponse: null,
+  setRequestAuthorization: null,
+  onAccessTokenAcquired: null,
```
onAccessTokenChange?

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+        refreshAccessTokenPromise = null;
+
+        tokens.accessToken = null;
+        tokens.refreshToken = null;
```
Move into clear() or reset()

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+        refreshAccessTokenPromise = null;
+
+        if (response.status !== STATUS_OK) {
+          throw new Error(ERROR_REFRESH_TOKEN_EXPIRED);
```
What if it's timeout or for example 204? You should probably handle error better and throw TOKEN_EXPIRED for only some specific 40? Maybe event call config.onRefreshTokenError?

dodsky replied:
Checked, I should only try to handle 401 and behave as normal fetch in all other cases, so if it throws, we rethrow. No magic.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+
+        if (config.onAccessTokenAcquired) {
+          config.onAccessTokenAcquired(token);
+        }
```
I would expect this block of code is moved in AccessTokenProvider into handleTokenChange or some other name. Should we name it like handleXY?

dodsky replied:
I think it should be called onAccessTokenChanged on config, and method handling it in config is a handler, and it is called handleAccessTokenChanged.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+
+function convertToRequest(args) {
+  const request = new Request(...args);
+  request.id = Date.now();
```
This cannot be `id`, it's not unique.

dodsky replied:
removed

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  return refreshAccessTokenPromise;
+}
+
+function convertToRequest(args) {
```
to `service`

dodsky replied:
removed

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  return request;
+}
+
+function shouldFetchAccessToken(request) {
```
part of AccessTokenProvider

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+    return false;
+  }
+
+  const requestAccessToken = parseBearer(request.headers.get('authorization'));
```
This should be moved into separate method because you use it twice in library and it's something important that should be encapsulated.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  return true;
+}
+
+function isAuthorized() {
```
AccessTokenProvider?

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  return !!tokens.refreshToken;
+}
+
+function fetchInterceptor(fetch, ...args) {
```
Part of Interceptor class `intercept`

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  const request = convertToRequest(args);
+
+  // check whether we should ignore this request
+  if (!isAuthorized() || !config.shouldIntercept(request)) {
```
Hide behind something like `AccessTokenProvider.shouldManage(...)

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+            })
+        })
+        .then(outerResolve)
+        .catch(outerReject);
```
Whole this tree should be compacted into simple call couple of methods that are encapsulating above behavior.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  }
+
+  // outer fetch promise
+  return new Promise((outerResolve, outerReject) => {
```
move into some method of Interceptor class

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+              .catch(outerReject);
+          }
+        });
+    }
```
This whole intercept method should be something like this, just suggestion.
```
Interceptor.intercept(...) {
  validate(...)
  createRequest(...) //spme service?
  
  this.authTokenProvider.shouldIntercept(...)) ..block
 
  const requestPromise = this.createRequestPromise(..) //returns promise

  this.authProvider.checktoken()
  if not valid:
     const renew = this.authProvider.renew()  
     // renew.then(requestPromise)...
  else
    requestPromise()  
}
```

dodsky replied:
Something similar :)

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+      }
+
+      // chain to refresh token promise (pending)
+      return refreshAccessTokenPromise
```
I would hide this `refreshAccessTokenPromise` in AccessTokenProvider

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+ * @param refreshToken
+ * @param accessToken
+ */
+export function authorize(refreshToken, accessToken) {
```
AccessTokenProvider?

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+ * Returns current authorization for fetch fetchInterceptor
+ * @returns {{accessToken: string, refreshToken: string}}
+ */
+export function getAuthorization() {
```
AccessTokenProvider?

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+/**
+ * Configures fetch token intercept
+ */
+export function configure(initConfig) {
```
Interceptor class and Interceptor class should pass part of config to AccessToken provider?

dodsky replied:
I'm currently passing the whole config to AccessTokenProvider until we settle on the design. It's easy to extract necessary methods in the end.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+  STATUS_OK,
+} from './const';
+
+let config = {
```
Document each config key

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+
+          return response;
+        })
+        .then(response => {
```
Please check usage of `.then` sometimes it's not necessary, I think once you encapsulate logic into methods that need for `.then` will diminish. You explained me concept of config methods that returns promise, and for that cases you don't have other options.

dodsky replied:
This whole part was redesigned into requestUnit idea as part of fetchInterceptor class and fetchWithRetry method.

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+    if (!tokens.accessToken) {
+      // check if we are alredy fetching it
+      if (!refreshAccessTokenPromise) {
+        // as a side-effect refreshTokenPromise is set
```
Don't like side-effect :)

dodsky replied:
Sideeffect is now part of accessTokenProvider :)

vedrani commented on src/index.js:
```diff
@@ -0,0 +1,250 @@
+            outerReject(error);
+          } else {
+            // if we fail to resolve refresh token, and it's not internal error
+            // just pass the request normally
```
This part is weird, I wouldn't expect it.

dodsky replied:
Here's how normal fetch behaves:
fetch(url).then(result).catch(error)
1. catch is called only on network errors (e.g. DNS unresolved)
2. then is called on all other responses, including 401

Initially I've solved this flow by throwing this exception, then capturing it in main fetch promise and resolving accordingly. In new version I did it without throwing exceptions.

vedrani reviewed (changes requested):
I commented this file a lot, but please use it only as suggestion. I think you did a great job :) I didn't review tests, I will look it in second iteration.  :1st_place_medal:

dodsky commented:
These are all great suggestions, I've already fixed some od these suggestions on the following commit, like encapsulating refresh token handling into separate class, and your ideja od interceptor correlates with my idea od interceptor provider. Also to enable promisifying any method od config I've converted a request into a smarter object which I've called 'requestUnit' which passes through interceptor promise pipeline and holds resolved values of all config functions. This effectively enables us to cashe values for resolved config methods and whole process acts as a middleware chain (got the ideja from node). :)

Commit:
Refactored initial solution into multiple files, including:
* fetchInterceptor - main class for intercepting
* accessTokenProvider - access token resolver
multiple services to increase readability as per PR comments

Commit:
Minor renames

Commit:
Added shouldFetch and onResponse
Fixed scope for fetch method

vedrani commented on test/helpers/tokenFormatter.spec.js:
```diff
@@ -0,0 +1,40 @@
+      expect(result).to.be.equal('token');
+    });
+  })
+});
```
new line :trollface:

dodsky replied:
![image](https://cloud.githubusercontent.com/assets/1985596/23251584/2f4ceec8-f9ae-11e6-9b10-7beaa40a48f0.png)

vedrani commented on package.json:
```diff
@@ -0,0 +1,59 @@
+    "access"
+  ],
+  "author": "Shoutem",
+  "license": "MIT",
```
BSD? custom?

dodsky replied:
Note, this should probably be updated on redux io repo too.

vedrani commented on package.json:
```diff
@@ -0,0 +1,59 @@
+  "author": "Shoutem",
+  "license": "MIT",
+  "bugs": {
+    "url": ""
```
link to github issues page?

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+  isResponseUnauthorized,
+} from './services/http';
+
+export class AccessTokenProvider {
```
Add nice jsdoc comment /** about this class, explaining it to others

vedrani replied:
Export **default**

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+  refresh() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      return Promise.resolve();
```
Should it reject?

dodsky replied:
If it rejects it will cause a chain reaction which will cause fetchIntercept chain to reject also, which in turn will reject an initial fetch request which is undesirable as we don't want to break the fetch spec which currently states that fetch will only reject on network errors and this is not a network error. 

But indeed it could be potentially problematic if you call `renew` without calling `authorize` first as it will silently ignore your `renew` requests. Two solutions here: silently fail with console error, or raise an exception which will be caught by fetchInterceptor which won't propagate it out of fetch call, but catch it, perhaps again log it, and then resolve as normal request would. I feel that the second solution only raises the handling to upper level without any flow control benefits as this can't be handled.

vedrani replied:
Looking from outside, I would want for my requests to fail with 40x error with equal behavior as in case that I don't use fetch-token-intercept and didn't add proper Auth header. Also it would be useful to see additional error or warning from fetch-token-intercept about need to call `authorize` with instruction to pass `refreshToken`.

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+  /**
+   * Refreshes current access token with provided refresh token
+   */
+  refresh() {
```
`renew` is better because refresh is in conflict with refreshToken

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+    this.fetch = fetch;
+
+    this.config = config;
+    this.refreshAccessTokenPromise = null;
```
`renew`

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+  resolveAccessToken(resolve, reject) {
+    return Promise.resolve(this.config.createAccessTokenRequest(this.tokens.refreshToken))
+      .then(this.fetchToken)
+      .then(this.handleFetchResolved)
```
handleFetchAccessTokenResponse

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  resolveAccessToken(resolve, reject) {
+    return Promise.resolve(this.config.createAccessTokenRequest(this.tokens.refreshToken))
```
move this nesting above `return`

```
const { refreshToken } = this.tokens;
const { createAccessTokenRequest } = this.config;

return Promise.resolve(createAccessTokenRequest(refreshToken))
...
```

vedrani commented on src/const.js:
```diff
@@ -0,0 +1,2 @@
+
```
remove line

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+
+  resolveAccessToken(resolve, reject) {
+    return Promise.resolve(this.config.createAccessTokenRequest(this.tokens.refreshToken))
+      .then(this.fetchToken)
```
fetchAccessToken

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+    return Promise.resolve(this.config.createAccessTokenRequest(this.tokens.refreshToken))
+      .then(this.fetchToken)
+      .then(this.handleFetchResolved)
+      .then(token => this.handleTokenResolved(token, resolve))
```
handleAccessToken

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+
+    if (isResponseUnauthorized(response)) {
+      this.clear();
+      return null;
```
Should you throw an exception?

dodsky replied:
Same reasoning as on the comment above. This is not exceptional behavior because it's not network related, we've successfully fetched a response, and it's status causes provider to reset itself. Promises waiting for it to resolve will carry on without a new token and fail with 401 which is expected. The gist here is 'try to reauthorize to get new token, if it fails, it should fail the same manner as without fetch interceptor'.

dodsky replied:
Again, we could throw, but fetchIntercept would have to handle it and resolve fetch requests normally. We can discuss this live.

vedrani replied:
I agree with your gist. I wasn't sure, so I rather ask to see what is your reasoning. Thanks.

vedrani commented on src/environment.js:
```diff
@@ -0,0 +1,62 @@
+  interceptors.push(new FetchInterceptor(env.fetch));
+
+  // monkey patch fetch
+  const fetchWrapper = fetch => (...args) => interceptors[0].intercept(...args);
```
It's a bit confusing to work with array[0]. What is idea behind multiple interceptors? Are they all going to intercept in parallel or in series requests? I would suggest that if you have `interceptors` and work with array you should support arrays everywhere instead of taking 0-index.

vedrani replied:
Below functions should also change, but we should first decide about interface for adding multiple interceptors. As I see for know we will have only one class `FetchInterceptor` but with different configurations (tokens, endpoints, callbacks, handlers). So maybe it could be `addInterceptor(key, config), authorizeInterceptor(key, ...args)`, ... What do you think?

dodsky replied:
As we discussed, we're at a point that this could be implemented easily when needed, and we don't really need it. I believe it goes to 'nice to have' list when we settle on implementation.

vedrani commented on src/environment.js:
```diff
@@ -0,0 +1,62 @@
+  configure,
+  authorize,
+  getAuthorization,
+}
```
Maybe this file could be called index.js. Currently, index.js is just re-exporting it.  Ok, it should also call `init()` :)

vedrani commented on src/environment.js:
```diff
@@ -0,0 +1,62 @@
+  return interceptors[0].clear();
+}
+
+export {
```
You should export direct on function signature.

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+import {
```
I this files with class export should be in PascalCase, also for AccessTokenProvider

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+import {
+  ERROR_INVALID_CONFIG,
+  ERROR_REFRESH_TOKEN_EXPIRED,
+} from '../lib/const';
```
Don't export from lib, also you are missing second variable from src/const

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+} from './services/http';
+import { AccessTokenProvider } from './accessTokenProvider';
+
+export class FetchInterceptor {
```
export **default**

vedrani commented on src/accessTokenProvider.js:
```diff
@@ -0,0 +1,111 @@
+
+    this.isAuthorized = this.isAuthorized.bind(this);
+    this.refresh = this.refresh.bind(this);
+    this.clear = this.clear.bind(this);
```
You divided public from private? Be free to add that comment. Maybe we should use some convention.

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+   * Configures fetch interceptor with given config object
+   * @param initConfig
+   */
+  configure(initConfig) {
```
rename arg to `config`

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+  fetchWithRetry(request, outerResolve, outerReject) {
+    return Promise.resolve(this.createRequestUnit(request))
+      .then(this.shouldIntercept)
+      // authorize request
```
Add this comment to each step in lifecycle

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+  shouldIntercept(requestUnit) {
+    const { request } = requestUnit;
+
+    return Promise.all([requestUnit, this.config.shouldIntercept(request)])
```
Do you still need Promise.all? Why not just ordinary promise that on then pass new requestUnit.

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+    return requestUnit;
+  }
+
+  authorizeRequest(requestUnit) {
```
move it above `shouldFetch` it's kinda nice to have order as in `fetchWithRetry`

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+    const { request } = requestUnit;
+
+    if (this.config.shouldFetch) {
+      return Promise.all([requestUnit, this.config.shouldFetch(request)])
```
Same

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+  authorizeRequest(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+
+    if (shouldIntercept) {
```
invert if

```
if (!shouldIntercept) {
  return requestUnit;
}
```

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+      const { request } = requestUnit;
+      const { accessToken } = this.accessTokenProvider.getAuthorization();
+
+      if (request && accessToken){
```
Same here. Please apply to all cases below.

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+      const { accessToken } = this.accessTokenProvider.getAuthorization();
+
+      if (request && accessToken){
+        return Promise.all([
```
Same here, please apply to all cases below.

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+      .then(this.fetchRequest)
+      .then(this.shouldInvalidateAccessToken)
+      .then(this.invalidateAccessToken)
+      .then(this.handleUnauthorizedRequest)
```
handleUnauthorizedRequest hides that you are renewing token. Maybe this could be more explicit, by rejecting and then in rejectHandler try to resolve problem?

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+          this.config.onResponse(response);
+        }
+
+        outerResolve(response);
```
This code in `then` block could be in handleResponse()?

dodsky replied:
The only reason that it isn't is because it needs access to methods for resolving fetch promise (outerResolve and outerReject are basically mapped like this fetch.then(outerResolve).then(outerReject)), but when I though about it, it totally makes sense to put references of outer fetch reject and resolve in request unit as any stage of pipeline could do

Commit:
Version bump

vedrani commented on src/fetchInterceptor.js:
```diff
@@ -0,0 +1,249 @@
+            shouldFetch: !!accessToken,
+          }))
+          .then(this.authorizeRequest)
+          .then(this.fetchRequest)
```
This block  in `if (isResponseUnauthorized(response)) {` is problematic, what happens if it fails twice? You only have one retry right? Should you  reuse fetchWithRetry and move request through flow once again?

vedrani commented on src/helpers/tokenFormatter.js:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  const bearerRegex = /^Bearer (.+)$/;
+  const matches = bearerRegex.exec(authorizationHeaderValue);
```
`string.match(regex)`

vedrani commented on test/fetchInterceptor.spec.js:
```diff
@@ -0,0 +1,345 @@
+        });
+    });
+
+    it('should propagate 401 for multiple requests', function (done) {
```
What happens here? Can we test what happens if one request gets multiple 401s?
1. Request is emitted with expired token
2. Response is 401
3. token is refreshed, request is reemitted
4. Response is 401
5. token is refreshed, request is reemitted
6. Response is 401
7. token is refreshed, request is reemitted
...

dodsky replied:
Wrote a test, it will retry as many times as it is defined on config with fetchRetryCount  property. (should stop after retry count is exceeded and resolve unauthorized)

vedrani reviewed (changes requested):
We should discuss some aspects of code.

Commit:
Fix invalid imports
Bump versoin

Commit:
Refactored flow logic with custom exceptions
Refactored tests
Minor renames and improvements

Commit:
Version bump

Commit:
Fixing invalid exception handling

vedrani commented on package.json:
```diff
@@ -0,0 +1,59 @@
+{
```
You don't ahve README file, add it. I know you don't have time to write everything. But maybe you could just add short description about what this lib solves and add link to somethnig like this:
`https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/`

vedrani commented on src/AccessTokenProvider.js:
```diff
@@ -0,0 +1,127 @@
+  renew() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      console.log('Please authorize provider before renewing or check shouldIntercept config.');
```
warning

vedrani commented on src/AccessTokenProvider.js:
```diff
@@ -0,0 +1,127 @@
+   * @param refreshToken
+   * @param accessToken
+   */
+  authorize(refreshToken, accessToken) {
```
Do you need to pass accessToken here?

dodsky replied:
I want to allow passing both refresh token and access token for two reasons:
1. perhaps you already have both stored externally and both are valid (why forcing a renew?)
2. RFC spec allows authorization response to return both https://tools.ietf.org/html/rfc6749#page-10

vedrani commented on src/AccessTokenProvider.js:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  handleAccessToken(token, resolve) {
+    this.tokens.accessToken = token;
```
this.tokens = {...this.tokens, accessToken}

vedrani commented on src/AccessTokenProvider.js:
```diff
@@ -0,0 +1,127 @@
+    return this.config.parseAccessToken(response);
+  }
+
+  handleAccessToken(token, resolve) {
```
accessToken

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+  }
+
+  /**
+   * Authorizes fetch interceptor with given renew token
```
You should use `refresh token` as name for token, sorry if I added confusion with `renew`. I only think `renew` word to be used for the process of getting new accessToken. But it doesn't need to be that way.

dodsky replied:
Oh... that's find/replace gone wild... Nice catch :) Fixed in class description also...

zrumenjak replied:
refresh?

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+  fetchRequest(requestUnit) {
+    const { shouldFetch } = requestUnit;
+
+    if (shouldFetch) {
```
invert `if`

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+    const { shouldIntercept } = requestUnit;
+    const { shouldInvalidateAccessToken } = this.config;
+
+    if (shouldIntercept && shouldInvalidateAccessToken) {
```
invert `if`

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+  invalidateAccessToken(requestUnit) {
+    const { shouldIntercept, shouldInvalidateAccessToken } = requestUnit;
+
+    if (shouldIntercept && shouldInvalidateAccessToken) {
```
same here

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+    }
+
+    if (this.config.onResponse) {
+      this.config.onResponse(response);
```
What if someone in onResponse function doesn't know how to handle `response` with `clone` is it going to destroy it before `fetchResolve`, should you pass response.clone here?

dodsky replied:
As this is the last callback it won't matter, but generally speaking, when accessing body you should clone first, because if anyone also has to read it after you it will fail I don't clone it first to avoid clone overhead when it's not necessary. Also, I've designed `onResponse` as fire and forget as we don't wait for events to finish during fetch intercept.

vedrani commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,327 @@
+    throw new Error(error);
+  }
+
+  isConfigValid() {
```
maybe put it higher in class to be closer to call

vedrani commented on src/services/RetryCountExceededException.js:
```diff
@@ -0,0 +1,15 @@
+  this.message = 'Retry count has been exceeded';
+  this.requestUnit = requestUnit;
+
+  // Use V8's native method if available, otherwise fallback
```
What we gain with this? It only for Chrome?

dodsky replied:
http://stackoverflow.com/a/27724419/1008529
`...with the enhancement that stack traces work on Firefox and other browsers. `

vedrani commented on src/services/RetryCountExceededException.js:
```diff
@@ -0,0 +1,15 @@
+
+RetryCountExceededException.prototype = Object.create(Error.prototype);
+RetryCountExceededException.prototype.name = "RetryCountExceededException";
+RetryCountExceededException.prototype.constructor = RetryCountExceededException;
```
new line.
Is there a ES6 way to do this? using `extends`

dodsky replied:
Babel 6.x does not correctly support extending core classes yet:
http://stackoverflow.com/questions/31089801/extending-error-in-javascript-with-es6-syntax 
https://github.com/babel/babel/issues/3083

there's a babel plugin https://www.npmjs.com/package/babel-plugin-transform-builtin-extend that enables extending from Error and Array. Don't know how we feel about this..

vedrani replied:
I would add this babel plugin, It's only overhead during transpiling but simplifies code.

vedrani reviewed (approved):
There is a couple of minor changes.

Commit:
Refactored exceptions to be more es6-like
Added readme
Minor renames in AccessTokenProvider and FetchInterceptor
Version bumped

zrumenjak commented on README.md:
```diff
@@ -0,0 +1,113 @@
+  //Event invoked when access token has changed
+  onAccessTokenChange: null,
+   
+  //Event invoked when response is resolved
```
You usually have a single space after // except in the examples above :)

zrumenjak commented on README.md:
````diff
@@ -0,0 +1,113 @@
+}
+```
+
+All required methods support returning a promise to enable reading of body.
````
Maybe `... return a promise to allow reading of the request body.`

zrumenjak commented on src/AccessTokenProvider.js:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  /**
+   * Returns current authorization for fetch fetchInterceptor
```
`fetch fetchInterceptor` is a bit confusing here...

zrumenjak commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,329 @@
+  resolveIntercept(resolve, reject, ...args) {
+    const request = new Request(...args);
+    const { accessToken } = this.accessTokenProvider.getAuthorization();
+    const requestUnit = this.createRequestUnit(request, resolve, reject);
```
What is a request unit? Maybe add a short description to explain it here.

dodsky replied:
Renamed to RequestContext, also added comments on createRequestContext method describing what it represents.

zrumenjak commented on src/FetchInterceptor.js:
```diff
@@ -0,0 +1,329 @@
+      .catch(this.handleUnauthorizedRequest);
+  }
+
+  createRequestUnit(request, fetchResolve, fetchReject) {
```
It wasn't clear to me what a request unit should be from its name. Maybe we should consider renaming it to something like `RequestContext` or possibly `RequestTask`.

zrumenjak reviewed (commented):
Looks awesome 👍
